File: binding/src/libqeventloop.cpp
#include <QDeadlineTimer>
#include <QEvent>
#include <QEventLoop>
#include <QEventLoopLocker>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QThread>
#include <qeventloop.h>
#include "libqeventloop.h"
#include "libqeventloop.hxx"

QEventLoop* QEventLoop_new() {
	 return new VirtualQEventLoop();
}

QEventLoop* QEventLoop_new2(QObject* parent) {
	 return new VirtualQEventLoop(parent);
}

libqt_string QEventLoop_Tr(const char* s) {
	QString _ret = QEventLoop::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QEventLoop_ProcessEvents(QEventLoop* self) {
	return self->processEvents();
}

void QEventLoop_ProcessEvents2(QEventLoop* self, int flags, int maximumTime) {
	self->processEvents(static_cast<QFlags<QEventLoop::ProcessEventsFlag>>(flags), maximumTime);
}

void QEventLoop_ProcessEvents3(QEventLoop* self, int flags, QDeadlineTimer* deadline) {
	self->processEvents(static_cast<QFlags<QEventLoop::ProcessEventsFlag>>(flags), *deadline);
}

int QEventLoop_Exec(QEventLoop* self) {
	return self->exec();
}

bool QEventLoop_IsRunning(const QEventLoop* self) {
	return self->isRunning();
}

void QEventLoop_WakeUp(QEventLoop* self) {
	self->wakeUp();
}

bool QEventLoop_Event(QEventLoop* self, QEvent* event) {
	return self->event(event);
}

void QEventLoop_Exit(QEventLoop* self) {
	self->exit();
}

void QEventLoop_Quit(QEventLoop* self) {
	self->quit();
}

libqt_string QEventLoop_Tr2(const char* s, const char* c) {
	QString _ret = QEventLoop::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QEventLoop_Tr3(const char* s, const char* c, int n) {
	QString _ret = QEventLoop::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QEventLoop_ProcessEvents1(QEventLoop* self, int flags) {
	return self->processEvents(static_cast<QFlags<QEventLoop::ProcessEventsFlag>>(flags));
}

int QEventLoop_Exec1(QEventLoop* self, int flags) {
	return self->exec(static_cast<QFlags<QEventLoop::ProcessEventsFlag>>(flags));
}

void QEventLoop_Exit1(QEventLoop* self, int returnCode) {
	self->exit(returnCode);
}

// Base class handler implementation
bool QEventLoop_QBaseEvent(QEventLoop* self, QEvent* event) {
	auto* vqeventloop = dynamic_cast<VirtualQEventLoop*>(self);
	if (vqeventloop && vqeventloop->isVirtualQEventLoop) {
vqeventloop->setQEventLoop_Event_IsBase(true);
	return vqeventloop->event(event);
}
}

// Auxiliary method to allow providing re-implementation
void QEventLoop_OnEvent(QEventLoop* self, intptr_t slot) {
	auto* vqeventloop = dynamic_cast<VirtualQEventLoop*>(self);
	if (vqeventloop && vqeventloop->isVirtualQEventLoop) {
vqeventloop->setQEventLoop_Event_Callback(reinterpret_cast<VirtualQEventLoop::QEventLoop_Event_Callback>(slot));
}
}

void QEventLoop_Delete(QEventLoop* self) {
    delete self;
}

QEventLoopLocker* QEventLoopLocker_new() {
	 return new QEventLoopLocker();
}

QEventLoopLocker* QEventLoopLocker_new2(QEventLoop* loop) {
	 return new QEventLoopLocker(loop);
}

QEventLoopLocker* QEventLoopLocker_new3(QThread* thread) {
	 return new QEventLoopLocker(thread);
}

void QEventLoopLocker_Swap(QEventLoopLocker* self, QEventLoopLocker* other) {
	self->swap(*other);
}

void QEventLoopLocker_Delete(QEventLoopLocker* self) {
    delete self;
}

File: binding/include/libqeventloop.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qeventloop.html
pub const qeventloop = struct {

    /// New constructs a new QEventLoop object.
    pub fn New() QtC.QEventLoop {
        return qtc.QEventLoop_new();
    }


    /// New2 constructs a new QEventLoop object.
    pub fn New2(parent: ?*anyopaque) QtC.QEventLoop {
        return qtc.QEventLoop_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QEventLoop_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qeventloop.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ProcessEvents(self: ?*anyopaque, ) bool {
        return qtc.QEventLoop_ProcessEvents(@ptrCast(self));
    }

    pub fn ProcessEvents2(self: ?*anyopaque, flags: i32, maximumTime: i32) void {
        qtc.QEventLoop_ProcessEvents2(@ptrCast(self), @intCast(flags), maximumTime);
    }

    pub fn ProcessEvents3(self: ?*anyopaque, flags: i32, deadline: QtC.QDeadlineTimer) void {
        qtc.QEventLoop_ProcessEvents3(@ptrCast(self), @intCast(flags), @ptrCast(deadline));
    }

    pub fn Exec(self: ?*anyopaque, ) i32 {
        return qtc.QEventLoop_Exec(@ptrCast(self));
    }

    pub fn IsRunning(self: ?*anyopaque, ) bool {
        return qtc.QEventLoop_IsRunning(@ptrCast(self));
    }

    pub fn WakeUp(self: ?*anyopaque, ) void {
        qtc.QEventLoop_WakeUp(@ptrCast(self));
    }

    pub fn Event(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QEventLoop_Event(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QEventLoop_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QEventLoop_QBaseEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn Exit(self: ?*anyopaque, ) void {
        qtc.QEventLoop_Exit(@ptrCast(self));
    }

    pub fn Quit(self: ?*anyopaque, ) void {
        qtc.QEventLoop_Quit(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QEventLoop_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qeventloop.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QEventLoop_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qeventloop.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ProcessEvents1(self: ?*anyopaque, flags: i32) bool {
        return qtc.QEventLoop_ProcessEvents1(@ptrCast(self), @intCast(flags));
    }

    pub fn Exec1(self: ?*anyopaque, flags: i32) i32 {
        return qtc.QEventLoop_Exec1(@ptrCast(self), @intCast(flags));
    }

    pub fn Exit1(self: ?*anyopaque, returnCode: i32) void {
        qtc.QEventLoop_Exit1(@ptrCast(self), returnCode);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QEventLoop_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qeventlooplocker.html
pub const qeventlooplocker = struct {

    /// New constructs a new QEventLoopLocker object.
    pub fn New() QtC.QEventLoopLocker {
        return qtc.QEventLoopLocker_new();
    }


    /// New2 constructs a new QEventLoopLocker object.
    pub fn New2(loop: ?*anyopaque) QtC.QEventLoopLocker {
        return qtc.QEventLoopLocker_new2(@ptrCast(loop));
    }


    /// New3 constructs a new QEventLoopLocker object.
    pub fn New3(thread: ?*anyopaque) QtC.QEventLoopLocker {
        return qtc.QEventLoopLocker_new3(@ptrCast(thread));
    }


    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QEventLoopLocker_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QEventLoopLocker_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/eventloop.html#types
pub const enums = struct {
    pub const ProcessEventsFlag = enum {
        pub const AllEvents: i32 = 0;
        pub const ExcludeUserInputEvents: i32 = 1;
        pub const ExcludeSocketNotifiers: i32 = 2;
        pub const WaitForMoreEvents: i32 = 4;
        pub const X11ExcludeTimers: i32 = 8;
        pub const EventLoopExec: i32 = 32;
        pub const DialogExec: i32 = 64;
        pub const ApplicationExec: i32 = 128;
    };

};
